src/mbgl/text/collision_tile.cpp

```diff
@@ -156,49 +148,53 @@ Box CollisionTile::getTreeBox(const Point<float>& anchor, const CollisionBox& bo
     };
 }
 
-std::vector<IndexedSubfeature> CollisionTile::queryRenderedSymbols(const GeometryCoordinates& queryGeometry, const float scale) {
-
+std::vector<IndexedSubfeature> CollisionTile::queryRenderedSymbols(const GeometryCoordinates& queryGeometry, float scale) {
     std::vector<IndexedSubfeature> result;
-    if (queryGeometry.empty()) return result;
-
-    std::unordered_map<std::string, std::unordered_set<std::size_t>> sourceLayerFeatures;
+    if (queryGeometry.empty() || (tree.empty() && ignoredTree.empty())) {
+        return result;
+    }
 
-    mapbox::geometry::multi_point<float> rotatedPoints {};
-    rotatedPoints.reserve(queryGeometry.size());
-    std::transform(queryGeometry.cbegin(), queryGeometry.cend(), std::back_inserter(rotatedPoints),
+    // Generate a rotated query box out of the original query geometry.
+    mapbox::geometry::multi_point<float> rotatedQuery {};
+    rotatedQuery.reserve(queryGeometry.size());
+    std::transform(queryGeometry.cbegin(), queryGeometry.cend(), std::back_inserter(rotatedQuery),
                    [&](const auto& c) { return util::matrixMultiply(rotationMatrix, convertPoint<float>(c)); });
-    const auto box = mapbox::geometry::envelope(rotatedPoints);
+    const auto box = mapbox::geometry::envelope(rotatedQuery);
 
-    const auto& anchor = box.min;
-    CollisionBox queryBox(anchor, 0, 0, box.max.x - box.min.x, box.max.y - box.min.y, scale);
-    auto predicates = bgi::intersects(getTreeBox(anchor, queryBox));
+    // Account for scaling when generating the query box.
+    const float x2 = (box.max.x - box.min.x) * scale;
+    const float y2 = (box.max.y - box.min.y) * scale;
+    const auto queryBox = CollisionBox { box.min, 0, 0, x2, y2, scale };
```

Comment:
I think I'm starting to understand.

The units of query box are correct. But `findPlacementScale` treats the query as a pyramid instead of a plain rectangular prism. What we want to make is a pyramid that covers the query box at `scale`. This means that if `scale != 1` then the base of the pyramid will be bigger. The base of the pyramid is the value we use to define the CollisionBox. So we need to calculate the size of the pyramid at `scale=1` so that at the current `scale` the pyramid is as big as the query box. The `* scale` calculates this correctly.